deployer/helm_upgrade_decision.py

```diff
@@ -0,0 +1,407 @@
+from rich.console import Console
+from rich.table import Table
+
+from utils import print_colour
+
+yaml = YAML(typ="safe", pure=True)
+
+# Determine if we are running a test or not. We set this env var to "test" in the
+# pyproject.toml file so it is set when the package is tested using pytest.
+test_env = os.getenv("RUN_ENV", False)
+
+
+def discover_modified_common_files(modified_paths: list):
+    """There are certain common files which, if modified, we should upgrade all hubs
+    and/or all clusters appropriately. These common files include the helm charts we
+    deploy, as well as the GitHub Actions and deployer package we use to deploy them.
+
+    Args:
+        modified_paths (list[str]): The list of files that have been added or modified
+            in a given GitHub Pull Request.
+
+    Returns:
+        upgrade_support_on_all_clusters (bool): Whether or not all clusters should have
+            their support chart upgraded since has changes
+        upgrade_all_hubs_on_all_clusters (bool): Whether or not all hubs on all clusters
+            should be upgraded since a core piece of infrastructure has changed
+    """
+    # If any of the following filepaths have changed, we should update all hubs on all clusters
+    common_filepaths = [
+        # Filepaths related to the deployer infrastructure
+        "deployer/*",
+        "requirements.txt",
+        # Filepaths related to GitHub Actions infrastructure
+        ".github/workflows/deploy-hubs.yaml",
+        ".github/actions/deploy/*",
+        # Filepaths related to helm chart infrastructure
+        "helm-charts/basehub/*",
+        "helm-charts/daskhub/*",
+    ]
+    # If this filepath has changes, we should update the support chart on all clusters
+    support_chart_filepath = "helm-charts/support/*"
+
+    # Discover if the support chart has been modified
+    upgrade_support_on_all_clusters = bool(
+        fnmatch.filter(modified_paths, support_chart_filepath)
+    )
+
+    # Discover if any common config has been modified
+    upgrade_all_hubs_on_all_clusters = False
+    for common_filepath_pattern in common_filepaths:
+        upgrade_all_hubs_on_all_clusters = bool(
+            fnmatch.filter(modified_paths, common_filepath_pattern)
+        )
+        if upgrade_all_hubs_on_all_clusters:
+            break
+
+    return upgrade_support_on_all_clusters, upgrade_all_hubs_on_all_clusters
+
+
+def get_unique_cluster_filepaths(added_or_modified_files: list):
+    """For a list of added and modified files, get the list of unique filepaths to
+    cluster folders containing added/modified files
+
+    Note: "cluster folders" are those that contain a cluster.yaml file.
+
+    Args:
+        added_or_modified_files (list[str]): A list of files that have been added or
+            modified in a GitHub Pull Request
+
+    Returns:
+        list[path obj]: A list of unique filepaths to cluster folders
+    """
+    # Get absolute paths
+    if test_env == "test":
```

Comment:
This introduces a global variable here. Given that this function is being called only from one place, i'd suggest either passing in `is_test` or `exclude_test` as a parameter to this function, or doing the filtering in the calling site itself. Global variables can become very difficult to remove after you add them.